grails-app/services/org/pih/warehouse/data/LocationDataService.groovy

```diff
@@ -33,32 +31,44 @@ class LocationDataService {
 
-            Location location = createOrUpdateLocation(params)
-
+            Location newLocation = Location.findByName(params.name)
+            Location location = new Location()
+            Location parentLocation = params.parentLocation ? Location.findByName(params.parentLocation) : null
             LocationGroup locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
+            LocationType locationType = params.locationType ? LocationType.findByName(params.locationType) : null
+
+            if (newLocation) {
+                location.errors.reject("Row ${index + 1}: '${params.name}' already exists in the system. Choose a unique name")
+            }
+
+            if (params.parentLocation && !parentLocation) {
+                location.errors.reject("Row ${index + 1}: Parent Location '${params.parentLocation}' does not exist")
+            }
+
             if (params.locationGroup && !locationGroup) {
-                command.errors.reject("Row ${index + 1}: Location group '${params.locationGroup}' does not exist")
+                location.errors.reject("Row ${index + 1}: Location Group '${params.locationGroup}' does not exist")
+            }
+
+            if (!locationType) {
+                location.errors.reject("Row ${index + 1}: Location Type '${params.locationType}' does not exist")
             }
 
-            if (!location.validate()) {
-                location.errors.each { BeanPropertyBindingResult error ->
-                    String errorMessage = messageSource.getMessage(error.fieldError, locale)
-                    command.errors.reject("Row ${index + 1}: ${errorMessage}")
-                }
+            if (locationType?.locationTypeCode in [LocationTypeCode.INTERNAL, LocationTypeCode.BIN_LOCATION, LocationTypeCode.ZONE] && !parentLocation) {
+                location.errors.reject("Row ${index + 1}: Location Type '${params.locationType}' requires a parent location")
             }
 
-            if (!location?.organization?.validate()) {
-                location?.organization?.errors?.each { BeanPropertyBindingResult error ->
-                    String errorMessage = messageSource.getMessage(error.fieldError, locale)
-                    command.errors.reject("Row ${index + 1}: ${errorMessage}")
-                }
+            if (!(locationType?.locationTypeCode in [LocationTypeCode.INTERNAL, LocationTypeCode.BIN_LOCATION, LocationTypeCode.ZONE]) && parentLocation) {
+                location.errors.reject("Row ${index + 1}: Cannot assign a parent location to a location of type '${params.locationType}'")
             }
 
+            location.errors.getAllErrors().each {
+                command.errors.reject(it.getCode())
+            }
         }
     }
 
     void importData(ImportDataCommand command) {
-        command.data.eachWithIndex { params, index ->
+        command.data.unique().eachWithIndex { params, index ->
```

Comment:
Can you explain the need for this? If someone has accidentally included the same row twice it feels like we want to warn them about that because there's probably a reason behind the data error. But maybe I'm missing something. In any case, let's also add a comment since it's not clear what's happening here.

Reply:
I just wonna to prevent errors while importing, because I thought that all voulnerabilities should be corrected during the validation, but you are right that in imported data could be two same rows that are not pointed during validation.